### PR TITLE
refactor[ux]: refactor preparser

### PR DESCRIPTION
**Pull request by tserg**
### What I did

This PR consolidates the pre-parsing process into a new PreParser object.

### Commit message

```
this PR refactors the pre-parsing routine to use a new `PreParser`
object. this will make it easier in the future to keep track of state
during pre-parsing.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ast/pre_parser.py [hunk 1]
```diff
@@ -158,184 +158,170 @@ def consume(self, token, result):
...
+        Parameters
+        ----------
+        code : str
+            The vyper source code to be re-formatted.
+        """
+        result: list[TokenInfo] = []
+        modification_offsets: dict[tuple[int, int], str] = {}
+        settings = Settings()
+        for_parser = ForParser(code)
+        hex_string_parser = HexStringParser()
```
Comment (tserg) [inline vyper/ast/pre_parser.py:197]: Renamed from `native_hex_parser`

File: vyper/ast/pre_parser.py [hunk 2]
```diff
@@ -158,184 +158,170 @@ def consume(self, token, result):
...
-                    if settings.compiler_version is not None:
-                        raise StructureException("compiler version specified twice!", start)
-                    compiler_version = contents.removeprefix("@version ").strip()
-                    validate_version_pragma(compiler_version, code, start)
-                    settings.compiler_version = compiler_version
-
-                if contents.startswith("pragma "):
-                    pragma = contents.removeprefix("pragma ").strip()
-                    if pragma.startswith("version "):
+    def parse(self, code: str):
```
Comment (tserg) [inline vyper/ast/pre_parser.py:173]: I think we can also make this the ctor: `def __init__(self, code: str)`

File: vyper/ast/pre_parser.py [hunk 3]
```diff
@@ -158,184 +158,172 @@ def consume(self, token, result):
...
     # Compilation settings based on the directives in the source code
     settings: Settings
     # A mapping of class names to their original class types.
     modification_offsets: dict[tuple[int, int], str]
     # A mapping of line/column offsets of `For` nodes to the annotation of the for loop target
     for_loop_annotations: dict[tuple[int, int], list[TokenInfo]]
-    # A list of line/column offsets of native hex literals
-    native_hex_literal_locations: list[tuple[int, int]]
+    # A list of line/column offsets of hex string literals
+    hex_string_locations: list[tuple[int, int]]
```
Comment (charles-cooper) [inline vyper/ast/pre_parser.py:169]: hex_literal_locations

**Review by charles-cooper — COMMENTED** [review]
it's nice, but there is a lot more nesting now

File: vyper/ast/pre_parser.py (hunk 3 above)
Comment (charles-cooper) [reply vyper/ast/pre_parser.py:169]: ehh, nvm

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**